### DISCUSSION TITLE
RFC - Amalgamations

### DISCUSSION THREAD

**Discussion by argush3** [bcgov/lear#2349] [General]
## Overview
This document is intended to help flesh out some of the design elements for amalgamation filings.

Specifically, we would like to establish a shared understanding of how to move forward with respect to the following items:
* How will the amalgamation information be represented from a data modelling perspective
* What actions will take place in the filer besides the creation of a new business
* Will an additional ledger item be created for historical amalgamating businesses to represent the amalgamation?
* What will the business response look like
* Does the UI have everything it needs from an API perspective to display amalgamation information for the amalgamating businesses and the amalgamated business?
  * ledger item
  * tombstone

## Db Model Updates

Add **amalgamating_business** and **amalgamation** table

![image](https://github.com/bcgov/lear/assets/6685223/1c1fd542-622d-4e9a-8657-0ca0facf8384)

## Filer Updates
* New business created per items defined in amalgamation filing json.  This will be similar to how business are created via an IA filing.
* New business will have state_filing_id populated with amalgamation filing id
* Amalgamating businesses will have following fields updated in **businesses** table
  * state = HISTORICAL
  * state_filing_id = amalgamation filing id (populated but not available in TING's business response)
  * dissolution_date = filing effective date
* Populate **amalgamating_businesses** table
* Populate **amalgamation** table

### Amalgamated Business
```http response

GET /businesses/<business_id>

No changes to current GET business endpoint.

{
  "business": {
    "identifier": "BC1531299",
    "legalName": "amalgamated business name",
    "legalType": "BC",
    "state": "ACTIVE",
    "hasCourtOrders": true
  }
}

```

### Amalgamating Business

Note that the state_filing is not populated as TINGS should not be able to access the TED's filing.
```http response
{
  "business": {
    "identifier": "BC1111111",
    "legalName": "amalgamating business name 1",
    "legalType": "BC",
    "state": "HISTORICAL",
    "dissolutionDate": "2010-06-09T07:00:00+00:00",
    "hasCourtOrders": false,
    "amalgamatedInto": {
      "amalgamationDate": "2010-06-09T07:00:00+00:00",
      "amalgamationType": "regular",
      "identifier": "BC1531299 ",
      "legalName": "amalgamated business name"
    }
  }
}

```

## UI
* Amalgamating businesses will have no additional ledger item for the amalgamation
* Amalgamation info returned in business response can be used for tombstone data specific to amalgamations

### Misc
* allowable actions can use amalgamation tables or state_filing_id to determine if a historical amalgamating business can be restored

**Comment by severinbeauvais** [comment]
I'm OK with having another table or 2 to hold amalgamation info, but the lack of a ledger entry requires quite a lot of extra info/complexity in the business response, which I think is undesirable.

**Reply by argush3** [reply]
I'm going to post an alternate version of business responses to just provide the info I see being used in legacy system.

The business summary can access the amalgamation tables directly behind the scenes.

**Reply by argush3** [reply]
I've added a "version 2" for the business responses.

Basically only amalgamating businesses have a change in business response.

Legacy system examples:

_TED_
![image](https://github.com/bcgov/lear/assets/6685223/932dd553-b7b3-426c-b75d-1423c68e2a0d)

_TING_
![image](https://github.com/bcgov/lear/assets/6685223/946638a6-2631-402a-adc9-a10092ef46b8)

**Reply by argush3** [reply]
yeah, I think I like v2 of the business response for TINGs if that is good enough for our needs.

**Reply by severinbeauvais** [reply]
I like v2 better as well.

>    "stateFiling": "https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/BC1531299/filings/147326",
>    "amalgamatedInto" ...

There's some redundancy here (they both provide the same view into what happened). ~And there's still a question whether this business (BC1111111) can fetch a filing indexed to the other business (BC1531299).~ _see other comment below_

If we definitely don't want a (hidden) final filing in the ledger then I think this is next best thing.

PS Is the ting still ACTIVE?

**Reply by argush3** [reply]
Updated TING to be HISTORICAL

**Comment by severinbeauvais** [comment]
Will the Legal API need a special authz for an amalgamating business to fetch the "state" filing using the amalgamated business' identifier?

**Reply by argush3** [reply]
No, doesn't look like the filings endpoint is locked down on a per business basis....for a single filing anyways

**Reply by severinbeauvais** [reply]
Oh. It feels like it should be, otherwise I can look at anyone's ledger.

**Reply by argush3** [reply]
yeah, feels like it should.  

Just not sure what if we open up a can of worms if we start locking it down.

**Reply by severinbeauvais** [reply]
Yeah, but what's worse? Worms or a security issue? 🤷‍♂️

**Reply by argush3** [reply]
Not saying we shouldn't lock it down.  Just envisioning the glass eating.

**Reply by severinbeauvais** [reply]
New ticket (for future):
https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/issues/gh/bcgov/entity/18968

**Comment by severinbeauvais** [comment]
Do you have an example of an amalgamating business' business response with the state filing id?

**Reply by argush3** [reply]
I don't have any real amalgamating business data.  

The business responses above are strictly theoretical examples.

**Reply by severinbeauvais** [reply]
Will an amalgamating business have both a state filing and the filing in the amalgamationInfo object?

**Reply by argush3** [reply]
Yes, and I just added a `state_filing` property to the amalgamating business response.

**Reply by severinbeauvais** [reply]
Nice.

Should the amalgamated business also have a state_filing -- does a corp show its IA as the state filing that caused it to exist?

**Reply by argush3** [reply]
I think it makes sense to include fill the state filing for creation of a businesses via IA, registration and amalgamation filing.

We don't do this yet for IA or registration filings yet though.  If we do there will need to be a data update exercise.

**Reply by severinbeauvais** [reply]
Sounds good to me. Curious what @thorwolpert thinks.

**Comment by severinbeauvais** [comment]
Also, I see a name mismatch between `dissolved_date` and `dissolutionDate`. Intentional?

**Reply by argush3** [reply]
Thanks and fixed.

Meant `dissolution_date`

**Comment by severinbeauvais** [comment]
Are we OK with an amalgamating business seeing the other amalgamating/ted business info? I thought Yui expressed an issue with that.

**Reply by argush3** [reply]
Well, info is pretty limited.  

Let me see if OrgBook provides all that info right now for amalgamating businesses.

**Reply by severinbeauvais** [reply]
Good point. There may be a business/legal requirement.

**Reply by severinbeauvais** [reply]
Not quite the same, but:

![image](https://github.com/bcgov/lear/assets/10002889/3bc558e4-91c4-47f3-a508-ea3744e78c9e)

We could always have a final ledger filing but not display it, same as bcgov/entity#16967.

**Reply by argush3** [reply]
So yeah there is limited info on the ledger and tombstones of TINGs and TEDs involved in amalgamations when I look into SOFI.

And Orgbook has minimal info too.

**Comment by severinbeauvais** [comment]
So, we can't resolve/close comments in here? Should we delete resolved ones?

**Reply by argush3** [reply]
what does hide do

**Comment by severinbeauvais** [comment]
> what does hide do

It hides the comment I hide, but not the whole thread.

**Reply by severinbeauvais** [reply]
It also looks like it creates a new thread 🤦

**Comment by severinbeauvais** [comment]
I still favour the idea of a "final filing" for  each amalgamating business (hidden from ledger with a flag). This:
- would be consistent with our [business state flow](https://github.com/bcgov/entity/blob/main/rfcs/rfc-business-states/business_states.png),
- wouldn't need one business to access another business' filings (ie, state filing), and
- wouldn't require "historical by amalgamation" to be a very special case.

**Reply by thorwolpert** [reply]
no ledger items should be hidden

**Reply by severinbeauvais** [reply]
I think Linda said that there should be no "final filing" for an amalgamating business. (Instead we might show an alert at the top of their business dashboard.) Are you OK with transitioning a TING business to historical without a state filing?

**Reply by argush3** [reply]
I think we would still populate the `state_filing_id` for amalgamating businesses.  

We may just not provide the state filing link for TING businesses as these businesses should not be able to access filings for the TED.

**Reply by severinbeauvais** [reply]
Hmm, ok, but we need to think about how a TING business will determine its historical reason. Up until now, a business looked at the state filing to determine the reason ([link](https://github.com/bcgov/business-filings-ui/blob/73cb7b0e3e454107f79d51fae8cdbac7265a3675/src/stores/rootStore.ts#L226)).

**Reply by argush3** [reply]
What about something like below?  With url hidden for amalgamating businesses.

I'm assuming you don't want to use the `amalgamatingInto` property from the business response as it breaks your generic pattern of using `stateFiling` property to determine the historical state reason.

![image](https://github.com/bcgov/lear/assets/6685223/2088ff14-7db9-46c1-a0e8-855ce3433875)

**Reply by severinbeauvais** [reply]
That would work.

Is the amalgamation date available anywhere? Re:

![image](https://github.com/bcgov/lear/assets/10002889/b654101c-ce8b-4eda-8f8e-2fc833cc655b)

**Reply by severinbeauvais** [reply]
Maybe that's the "date of dissolution"?

**Reply by argush3** [reply]
>Is the amalgamation date available anywhere?

So are you saying you don't want the `amalgamationInto` block in the business response?  That blocks contains the amalgamation date and a few other things.

**Reply by severinbeauvais** [reply]
Oops, I missed that.

Looks like the `amalmagationInto` block has all the info the UI needs for a historical reason.

Maybe the state filing is redundant then? Especially if the TING cannot open it. What do you think - could it be used in the future? Could it be used to cross-reference filings between businesses (kind of rebuilding the amalgamation references)?

**Reply by argush3** [reply]
I think we go back to original suggestion of not showing the state filing then.

The state_filing will still be set in the business table and still be used by the BE in addition to the new amalgamation tables.

**Reply by severinbeauvais** [reply]
Yes. Sorry about the run-around on this one 😅

**Comment by argush3** [comment]
@severinbeauvais  @thorwolpert I've updated the RFC to an end state that I think we can move forward with.  

If anyone has any objections, stop me now!

**Reply by severinbeauvais** [reply]
I think there are some tiny details to work out but I think this will work.

In the end: amalgamations will require a special case to handle them in the UI (ie, separate from other historical businesses).

**Comment by severinbeauvais** [comment]
> New business will have state_filing_id populated with amalgamation filing id

~Is that what we do for IAs also?~ _I already asked this question above, though there's no resolution on this yet._

**Reply by argush3** [reply]
No, we don't it currently for IAs or registrations but as per our previous discussion, it makes sense I think.  

My thought is we add it for amalgamations(TEDs) and have future tickets to update filer for IAs/registrations as well as backfill the data.

**Reply by severinbeauvais** [reply]
For IAs, Regs, and the TED 👍

**Comment by severinbeauvais** [comment]
Did we ever decide whether to show an alert on the historical TINGs? The UI designs don't show one, and I'm not sure it's needed, but it could be used to approximate what COLIN currently shows (or we could add a tooltip to the HISTORICAL badge - I'll ask Yui).
